### PR TITLE
Add the feature for comparing two arbitrary commits

### LINKED ISSUE #985: Add ability to display the diff of any two commits in the file history panel
Select two arbitrary commits and view their diff (`Select for compare` & `Compare with selected` from VSCode)

_Originally posted by @navn-r in https://github.com/jupyterlab/jupyterlab-git/issues/980#issuecomment-886121034_

### PR DESCRIPTION
**Pull request by quachtridat**
## Summary

This PR resolves #985 by implementing the necessary components and logic that will allow users to select two commits in the *History* tab and see the Git diff for every changed file.

### Backend

https://github.com/jupyterlab/jupyterlab-git/pull/1108/commits/5cee9bfec6dafe0dfcc826cc83b34e2b61bc3b8b

### Frontend

Each entry in `History` tab now has two buttons, each with an icon of a document containing an arrow pointing right/left. Clicking one of these will make the commit comparison box pop up, and the diff will be evaluated when there are two commits selected for comparison. Stylings are similar to how the `Changed` section for a commit entry in `History` is styled.

## Caveats

The commit comparison box can be collapsed, but cannot be resized for now.

## Demos

image
*The commit that is __selected for compare__ has a little red color on the left side.*

image
*The commit that is to be __compared with selected__ has a little green color on the left side.*

image
*The comparison results are shown in the commit comparison box at the bottom of the `History` tab.*

image
*The diff for the file `README.md` is shown when the diff button for that file is clicked.*

image
*The commit comparison box when it is collapsed.*

## Notes

This PR may need more tests.

## Linked issues

#985

### CODE REVIEW

File: src/style/PastCommitNode.ts [hunk 1]
```diff
@@ -83,3 +83,11 @@ export const singleFileCommitClass = style({
     }
   }
 });
+
+export const commitCompareLhsNodeClass = style({
+  borderLeft: '3px solid darkred'
```
Comment (fcollonval) [inline src/style/PastCommitNode.ts:88]: Could you try with 
```suggestion
  borderLeft: '3px solid var(--jp-git-diff-deleted-color)'
```

File: src/style/PastCommitNode.ts [hunk 2]
```diff
@@ -83,3 +83,11 @@ export const singleFileCommitClass = style({
...
+});
+
+export const commitCompareRhsNodeClass = style({
+  borderLeft: '3px solid darkgreen'
```
Comment (fcollonval) [inline src/style/PastCommitNode.ts:92]:
```suggestion
  borderLeft: '3px solid var(--jp-git-diff-added-color)'
```

**Review by fcollonval — CHANGES_REQUESTED** [review]
Thanks @quachtridat and @ZeshanFayyaz 

I mainly have suggestions for a better wording (using reference/challenger instead of lhs/rhs) and a small rework to avoid storing twice the same information in the panel state.

File: jupyterlab_git/git.py [hunk 3]
```diff
@@ -630,13 +636,14 @@ async def diff(self, path):
...
         for line in line_array:
             linesplit = line.split()
-            result.append(
-                {
-                    "insertions": linesplit[0],
-                    "deletions": linesplit[1],
-                    "filename": linesplit[2],
-                }
-            )
+            if len(linesplit) == 3:
```
[the remark this answers is not in the document]
Comment (quachtridat) [reply jupyterlab_git/git.py:639]: > Why is this test required? In which case are those information missing?

I have not manually tested without this, but in early development of this feature I got an error without that line (and it was `IndexError` if I remember correctly), and no error with that line.

[the remark this answers is not in the document]
Comment (fcollonval) [reply jupyterlab_git/git.py:639]: thanks @quachtridat would you mind adding a else condition that log at debug level the _broken_ line?

File: src/tokens.ts [hunk 4]
```diff
@@ -1000,6 +1031,25 @@ export namespace Git {
...
+   */
+  export interface ICommitComparison {
+    /**
+     * The commit to compare against
+     */
+    lhs: ISingleCommitInfo;
+    /**
+     * The commit to compare
+     */
+    rhs: ISingleCommitInfo;
```
[the remark this answers is not in the document]
Comment (quachtridat) [reply src/tokens.ts:1045]: That means we should just pass `referenceCommit`, `challengerCommit` and `comparison` one-by-one as props for `CommitComparisonBox`?

File: src/components/CommitComparisonBox.tsx [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+  comparison: Git.ICommitComparison | null;
+  header: string;
+  model: IGitExtension;
+  trans: TranslationBundle;
+  onCancel?: (event: React.MouseEvent<HTMLElement, MouseEvent>) => void;
+  onOpenDiff?: (
+    filePath: string,
+    isText: boolean,
+    previousFilePath?: string
+  ) => (event: React.MouseEvent<HTMLLIElement, MouseEvent>) => void;
```
Comment (quachtridat) [inline src/components/CommitComparisonBox.tsx:228]: TODO: docstring.
@ZeshanFayyaz

Comment (quachtridat) [inline src/components/CommitComparisonBox.tsx:228]: TODO: docstring.
@ZeshanFayyaz

File: jupyterlab_git/git.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (quachtridat) [reply jupyterlab_git/git.py:639]: > thanks @quachtridat would you mind adding a else condition that log at debug level the _broken_ line?

Sure!
@ZeshanFayyaz we might as well investigate this.

File: jupyterlab_git/handlers.py [hunk 6]
```diff
@@ -258,7 +258,16 @@ async def post(self, path: str = ""):
         POST request handler, fetches differences between commits & current working
```
Comment (fcollonval) [inline jupyterlab_git/handlers.py:258]:
```suggestion
        POST request handler, fetches differences between two states
```

**Review by fcollonval — APPROVED** [review]
Thanks a lot @quachtridat and @ZeshanFayyaz 

This is a great addition

**Comment by fcollonval** [comment]
@all-contributors please add quachtridat for code

**Comment by fcollonval** [comment]
@all-contributors please add @ZeshanFayyaz for code

**Pull request merged**